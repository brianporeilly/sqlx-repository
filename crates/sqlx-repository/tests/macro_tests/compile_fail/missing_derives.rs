//! Test that repository derive fails when required derives are missing

use sqlx_repository::prelude::*;

// Missing Debug, Clone, Serialize, Deserialize, sqlx::FromRow
#[derive(Repository)]
#[repository(table = "users")]
pub struct User {
    pub id: i32,
    pub name: String,
    pub email: String,
}

fn main() {}