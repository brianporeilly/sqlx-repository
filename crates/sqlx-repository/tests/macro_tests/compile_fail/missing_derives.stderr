error[E0277]: the trait bound `for<'r> User: FromRow<'r, PgRow>` is not satisfied
 --> tests/macro_tests/compile_fail/missing_derives.rs:6:10
  |
6 | #[derive(Repository)]
  |          ^^^^^^^^^^ the trait `for<'r> FromRow<'r, PgRow>` is not implemented for `User`
  |
  = help: the following other types implement trait `FromRow<'r, R>`:
            ()
            (T1, T2)
            (T1, T2, T3)
            (T1, T2, T3, T4)
            (T1, T2, T3, T4, T5)
            (T1, T2, T3, T4, T5, T6)
            (T1, T2, T3, T4, T5, T6, T7)
            (T1, T2, T3, T4, T5, T6, T7, T8)
          and $N others
note: required by a bound in `sqlx_repository::Repository`
 --> src/repository.rs
  |
  | pub trait Repository<T>: Send + Sync
  |           ---------- required by a bound in this trait
  | where
  |     T: for<'r> FromRow<'r, sqlx::postgres::PgRow> + Send + Unpin,
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Repository`
  = note: this error originates in the derive macro `Repository` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `for<'r> User: FromRow<'r, PgRow>` is not satisfied
 --> tests/macro_tests/compile_fail/missing_derives.rs:6:10
  |
6 | #[derive(Repository)]
  |          ^^^^^^^^^^ the trait `for<'r> FromRow<'r, PgRow>` is not implemented for `User`
  |
  = help: the following other types implement trait `FromRow<'r, R>`:
            ()
            (T1, T2)
            (T1, T2, T3)
            (T1, T2, T3, T4)
            (T1, T2, T3, T4, T5)
            (T1, T2, T3, T4, T5, T6)
            (T1, T2, T3, T4, T5, T6, T7)
            (T1, T2, T3, T4, T5, T6, T7, T8)
          and $N others
note: required by a bound in `soft_delete_enabled`
 --> src/repository.rs
  |
  |     T: for<'r> FromRow<'r, sqlx::postgres::PgRow> + Send + Unpin,
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Repository::soft_delete_enabled`
...
  |     fn soft_delete_enabled() -> bool {
  |        ------------------- required by a bound in this associated function
  = note: this error originates in the derive macro `Repository` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `for<'r> User: FromRow<'r, PgRow>` is not satisfied
 --> tests/macro_tests/compile_fail/missing_derives.rs:6:10
  |
6 | #[derive(Repository)]
  |          ^^^^^^^^^^ the trait `for<'r> FromRow<'r, PgRow>` is not implemented for `User`
  |
  = help: the following other types implement trait `FromRow<'r, R>`:
            ()
            (T1, T2)
            (T1, T2, T3)
            (T1, T2, T3, T4)
            (T1, T2, T3, T4, T5)
            (T1, T2, T3, T4, T5, T6)
            (T1, T2, T3, T4, T5, T6, T7)
            (T1, T2, T3, T4, T5, T6, T7, T8)
          and $N others
note: required by a bound in `find_by_id`
 --> src/repository.rs
  |
  |     T: for<'r> FromRow<'r, sqlx::postgres::PgRow> + Send + Unpin,
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Repository::find_by_id`
...
  |     async fn find_by_id(&self, id: i32) -> RepositoryResult<Option<T>> {
  |              ---------- required by a bound in this associated function
  = note: this error originates in the derive macro `Repository` (in Nightly builds, run with -Z macro-backtrace for more info)
